### PR TITLE
Higher Nim in Action coverage plus regression fix.

### PR DESCRIPTION
**Pull request by dom96**
* MD5 hashes of Nim in Action tests are now checked to prevent accidental modification.
* ``strutils`` module now exports unicode.toLower/toUpper.
* Added smaller code samples from Chapters 1-3.

### CODE REVIEW

File: tests/testament/categories.nim [hunk 1]
```diff
@@ -278,8 +287,34 @@ proc testNimInAction(r: var TResults, cat: Category, options: string) =
...
+  # whoever edits these hashes without dom96's permission, j/k. But please only
+  # edit when making a conscious breaking change, also please try to make your
+  # commit message clear so I can easily compile an errata later.
+  var testHashes: seq[string] = @[]
+
+  for test in tests:
+    testHashes.add(getMD5(readFile("tests" / test.addFileExt("nim")).string))
+
+  let refHashes = @[
+    "51afdfa84b3ca3d810809d6c4e5037ba", "30f07e4cd5eaec981f67868d4e91cfcf",
```
Comment (Araq) [inline tests/testament/categories.nim]: Is this serious? Cause it sucks.

Comment (dom96) [reply tests/testament/categories.nim]: It is serious. Why does it suck?

**Comment by dom96** [comment]
Is ``nimvm`` not the right define? Why is it still failing with that error?

File: tests/testament/categories.nim [hunk 2]
```diff
@@ -278,8 +287,34 @@ proc testNimInAction(r: var TResults, cat: Category, options: string) =
...
+  # Verify that the files have not been modified. Death shall fall upon
+  # whoever edits these hashes without dom96's permission, j/k. But please only
+  # edit when making a conscious breaking change, also please try to make your
+  # commit message clear so I can easily compile an errata later.
+  var testHashes: seq[string] = @[]
+
+  for test in tests:
+    testHashes.add(getMD5(readFile("tests" / test.addFileExt("nim")).string))
+
+  let refHashes = @[
```
Comment (Araq) [inline tests/testament/categories.nim]: Make that a ``const`` array.

File: lib/pure/strutils.nim [hunk 3]
```diff
@@ -17,6 +17,10 @@ import parseutils
 from math import pow, round, floor, log10
 from algorithm import reverse
 
+when not defined(nimvm):
```
Comment (Araq) [inline lib/pure/strutils.nim]: Use the new ``defined(nimVmExportFixed)`` switch here.

**Pull request merged**